File: baekjoon/bfs/1963.js
function getPrimes() {
  const count = Array(10000).fill(false);
  const primes = Array(10000).fill(false);

  for (let i = 2; i < 10000; i++) {
    if (count[i] === false && i > 1000) primes[i] = true;

    for (let j = i; j < 10000; j += i) {
      count[j] = true;
    }
  }

  return primes;
}

function getInput() {
  const fs = require('fs');
  const input = fs
    .readFileSync('./baekjoon/bfs/1963.txt')
    .toString()
    .trim()
    .split('\n')
    .map(v => v.split(' ').map(Number));

  return [input[0][0], input.slice(1)];
}

function replaceAt(string, index, replacement) {
  return string.substring(0, index) + replacement + string.substring(index + 1);
}

function solution(primes, [start, target]) {
  let optStep = -1;

  const notVisited = [...primes];
  const queue = [[0, start]];
  notVisited[start] = false;

  while (queue.length > 0) {
    const [curStep, curNum] = queue.shift();

    if (curNum === target) {
      optStep = curStep;
      break;
    }

    const curNumStr = curNum.toString();

    for (let i = 0; i < 10; i++) {
      for (let j = 0; j < 4; j++) {
        const nextNum = Number(replaceAt(curNumStr, j, i.toString()));

        if (!primes[nextNum]) continue;

        if (notVisited[nextNum]) {
          queue.push([curStep + 1, nextNum]);
          notVisited[nextNum] = false;
        }
      }
    }
  }

  return optStep;
}

function main() {
  const primes = getPrimes();
  const [T, cases] = getInput();

  let ret = [];
  for (let tc = 0; tc < T; tc++) {
    const ans = solution(primes, cases[tc]);
    ans >= 0 ? ret.push(ans) : ret.push('impossible');
  }

  console.log(ret.join('\n'));
}

if (require.main === module) main();

File: baekjoon/bfs/1963.py
from collections import deque
import math
import sys
from typing import List, Set

input = sys.stdin.readline


def get_primes() -> Set[int]:
    n = 9999  # 2부터 1,000까지의 모든 수에 대하여 소수 판별
    array = [True for i in range(n + 1)]  # 처음엔 모든 수가 소수(True)인 것으로 초기화

    # 에라토스테네스의 체 알고리즘
    for i in range(2, int(math.sqrt(n)) + 1):  # 2부터 n의 제곱근까지의 모든 수를 확인하며
        if array[i] == True:  # i가 소수인 경우 (남은 수인 경우)
            # i를 제외한 i의 모든 배수를 지우기
            j = 2
            while i * j <= n:
                array[i * j] = False
                j += 1

    primes = set()
    for i, v in enumerate(array):
        if v:
            primes.add(i)

    return primes


def solution(src: int, target: int, primes: Set[int]):
    result = bfs(src, target, primes)

    if result < 0:
        print("Impossible")
    else:
        print(result)


def bfs(start: int, target: int, primes: Set[int]) -> int:
    queue = deque([(start, 0)])
    visited = set([start])
    d = [(0, True), (0, False), (1, True), (1, False), (2, True), (2, False), (3, True), (3, False)]

    while queue:
        cur_prime, cur_repeat = queue.popleft()

        if cur_prime == target:
            return cur_repeat

        for i in range(8):
            exp_num, is_add = d[i]
            op_num = pow(10, exp_num)
            before = cur_prime // pow(10, exp_num + 1)

            for j in range(1, 10):
                if is_add:
                    next_num = cur_prime + (op_num * j)
                else:
                    next_num = cur_prime - (op_num * j)

                after = next_num // pow(10, exp_num + 1)

                if after != before:
                    continue

                if not (1000 <= next_num < 10000):
                    continue

                if next_num in visited:
                    continue

                if next_num not in primes:
                    continue

                visited.add(next_num)
                queue.append((next_num, cur_repeat + 1))

    return -1


if __name__ == "__main__":
    case_num = int(input())
    primes = get_primes()

    for _ in range(case_num):
        src, target = list(map(int, input().split()))
        solution(src, target, primes)
